### PR TITLE
Slang 0.18.3

### PR DESCRIPTION
**Pull request by Janther**
adding support for slang v0.18.0 and WebAssembly.

Main changes:

  - adapting to small changes in the slang API
  - dropping (temporarily hopefully) support for UMD
  - bundling an esm package
  - removing all the use of the slang AST that is not necessary

### CODE REVIEW

File: src/slangPrinter.ts [hunk 1]
```diff
@@ -82,8 +62,6 @@ function genericPrint(
   options: ParserOptions<AstNode>,
   print: PrintFunction
 ): Doc {
```
Comment (OmarTawfik) [inline src/slangPrinter.ts:64]: If this PR is explicitly dropping support for v2, I wonder if this should be kept, specifying `3.0.0` instead?

File: src/printer.js [hunk 2]
```diff
@@ -23,18 +23,8 @@ const warnDeprecation = once(() => {
   return true;
 });
```
Comment (OmarTawfik) [inline src/printer.js:24]: If this PR is explicitly dropping support for v2, I wonder if this should be kept, specifying `3.0.0` instead?

File: tests/config/get-plugins.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+async function getPluginsInternal() {
+  if (!process.env.TEST_STANDALONE) {
+    return [path.join(__dirname, "../../src/index.ts")];
+  }
+
+  const require = createRequire(import.meta.url);
+  const prettier = await getPrettier();
+
+  const prettierPath = path.dirname(require.resolve("prettier"));
+  const pluginPrefix = satisfies(prettier.version, "^2.3.0")
```
Comment (OmarTawfik) [inline tests/config/get-plugins.js:18]: If this PR is explicitly dropping support for v2, I wonder if this should be kept, specifying `3.0.0` instead?

File: src/printer.js (hunk 2 above)
Comment (OmarTawfik) [reply src/printer.js:24]: Same question for:

- `src/common/backward-compatibility.js`
- `src/slangPrinter.ts`
- `tests/config/get-plugins.js`

File: src/slangPrinter.ts [hunk 4]
```diff
@@ -1,29 +1,9 @@
...
-      value = factory();
-    }
-    return value;
-  };
-}
-
-const prettierVersionCheck = once((): boolean => {
-  if (!prettierVersionSatisfies('>=2.3.0')) {
-    throw new Error(
-      'The version of prettier in your node-modules does not satisfy the required ">=2.3.0" constraint. Please update the version of Prettier.'
```
Comment (OmarTawfik) [inline src/slangPrinter.ts:22]: If this PR is explicitly dropping support for v2, I wonder if this should be kept, specifying `3.0.0` instead?

File: tests/config/get-plugins.js (hunk 3 above)
Comment (OmarTawfik) [inline tests/config/get-plugins.js:18]: remaining v2 support. I wonder if it should be removed?

Same question for `src/common/backward-compatibility.js`

File: tests/unit/slang-utils/create-parser.test.js [hunk 5]
```diff
@@ -90,14 +90,14 @@ describe('inferLanguage', function () {
...
     test(description, function () {
-      const inferredLanguage = inferLanguage(source);
-      expect(inferredLanguage.version).toEqual(version);
+      const parser = createParser(source);
+      expect(parser.version).toEqual(version);
     });
   }
 
-  test('should throw an error if there are incompatible ranges', function () {
+  test.skip('should throw an error if there are incompatible ranges', function () {
```
Comment (OmarTawfik) [inline tests/unit/slang-utils/create-parser.test.js:98]: this test is skipped. not sure if that was intended?

File: src/slang-utils/create-parser.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    supportedVersions
+  );
+
+  return satisfyingVersions.length > 0
+    ? Parser.create(satisfyingVersions[satisfyingVersions.length - 1])
+    : Parser.create(supportedVersions[supportedVersions.length - 1]);
+}
+
+function tryToCollectPragmas(text: string, version: string): string[] {
+  const language = Parser.create(version);
```
Comment (OmarTawfik) [inline src/slang-utils/create-parser.ts:61]: suggestion to remove the need for the extra dependency, and to be able to handle things like newlines/other corner cases:

```ts
const fileCursor = parser.parse(NonterminalKind.SourceUnit, source).createTreeCursor();
const values = [];

while (fileCursor.goToNextNonterminalWithKind(NonterminalKind.VersionExpressionSets)) {
  const versionCursor = fileCursor.spawn();
  let value = "";

  while (versionCursor.goToNextTerminal()) {
    const terminal = versionCursor.node.asTerminalNode()!;
    if (TerminalKindExtensions.isTrivia(terminal.kind)) {
      value += " "; // to break up any possible consecutive identifiers
    } else {
      value += terminal.unparse();
    }
  }

  values.push(value);
}
```

File: package.json [hunk 7]
```diff
@@ -1,18 +1,27 @@
 {
```
Comment (OmarTawfik) [inline package.json:1]: I wonder if we have integration tests that run this inside a browser?
I see that the `README.md` says `_Disabled during v2.0.0-beta_`. Is this correct?

**Review by OmarTawfik — COMMENTED** [review]
Left a few suggestions/question. Thank you! 🚀🚀

File: src/printer.js (hunk 2 above)
Comment (Janther) [reply src/printer.js:24]: prettier 2.2.X would still load the plugin and fail miserably when using it, so we needed the check for `>=2.3.0` to show the reasons why it was failing.
Now prettier 2.X.X won't even load the plugin so the code will never reach this code.
there is a check in `peerDependencies` in `package.json` and I'll add this info on the readme.

File: tests/config/get-plugins.js (hunk 3 above)
Comment (Janther) [reply tests/config/get-plugins.js:18]: yes. thanks for finding this.

File: tests/unit/slang-utils/create-parser.test.js (hunk 5 above)
Comment (Janther) [reply tests/unit/slang-utils/create-parser.test.js:98]: It's intended to be skipped for now.

before we were joining pragma versions into a single string and it was straight forward to find an inconsistency, but as you pointed out, there are some edge cases where this approach leads to an error. Now we evaluate each pragma version against slang's supported versions, and now an incompatibility in the pragma versions is no different from an eager developer chasing to use the latest solidity release which might not be supported by slang just yet but with a syntax fully supported by slang.

so we might choose to remove this test as it might not be a responsibility of `prettier-solidity` or we will have to add extra steps in the version inference to double check for the validity of the pragmas.

File: package.json (hunk 7 above)
Comment (Janther) [reply package.json:1]: We were running all of our tests in a virtual empty javascript context. But had to drop it because it wouldn't load `wasm` files and that configuration was design for `commonjs`.
I'll research adding an integration test with a browser. `karma` seems to be deprecated now.

I'll update the Readme

File: src/slang-utils/create-parser.ts (hunk 6 above)
Comment (Janther) [reply src/slang-utils/create-parser.ts:61]: I tried using this and tests aren't passing. I suspect is because we might be moving through an AST with errors when the syntax doesn't match the parser's version.

File: package.json (hunk 7 above)
Comment (OmarTawfik) [reply package.json:1]: I've used `puppeteer` before for CI testing, and IIRC, you can load an `.html` in-memory and just query the elements there.

File: src/slang-utils/create-parser.ts (hunk 6 above)
Comment (OmarTawfik) [reply src/slang-utils/create-parser.ts:61]: That syntax should be consistent across all versions. I wonder if you have a commit # I can try out?
Happy to help if I can debug the failure..

**Review by OmarTawfik — APPROVED** [review]
